src/awkward/_v2/index.py

```diff
@@ -1,79 +1,84 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+from __future__ import absolute_import
+
 import awkward as ak
 
+np = ak.nplike.NumpyMetadata.instance()
+
+
+class Index(object):
+    _dtype_to_form = {
+        np.dtype(np.int8): "i8",
+        np.dtype(np.uint8): "u8",
+        np.dtype(np.int32): "i32",
+        np.dtype(np.uint32): "u32",
+        np.dtype(np.int64): "i64",
+    }
 
-class Index:
     def __init__(self, data):
         self._nplike = ak.nplike.of(data)
+
         self._data = self._nplike.asarray(data, order="C")
+        if len(self._data.shape) != 1:
+            raise TypeError("an actual error string")
+
         self._T = self._data.dtype
-        if self._data == None:
-            self._len = 0
-        else:
-            self._len = len(self.data)
-        if self._T not in (
-            self._nplike.int8,
-            self._nplike.uint8,
-            self._nplike.int32,
-            self._nplike.uint32,
-            self._nplike.int64,
-            object,
-        ):
-            raise TypeError
-
-    # the default for this is float64 so maybe type should also be mentioned
-    def __zeros__(length, nplike):
-        return nplike.zeros(length)
-
-    # the default for this is float64 so maybe type should also be mentioned
-    def __empty__(length, nplike):
-        return nplike.empty(length)
-
-    def __data__(self):
+        if self._T not in self._dtype_to_form:
+            raise TypeError("an actual error string")
+
+    @classmethod
+    def zeros(cls, length, nplike, dtype):
+        return nplike.zeros(length, dtype=dtype)
+
+    @classmethod
+    def empty(cls, length, nplike, dtype):
+        return nplike.empty(length, dtype=dtype)
```

Comment:
Alternate constructors should be `@classmethods`, which take the class (`cls`) as the first argument (instead of `self` like a normal method or nothing like a `@staticmethod`).

You are right that a data type must be specified. I forgot about that!